### PR TITLE
Bnb/sup3r mini model

### PR DESCRIPTION
**Pull request by bnb32**
Added recursive layer initialization in `HiddenLayers` and `Sup3rObsModel` which can be built from nested hidden layers in a config, and also ingest multiple obs features / exo features.

### CODE REVIEW

File: phygnn/layers/custom_layers.py [hunk 1]
```diff
@@ -1527,23 +1539,42 @@ def call(self, x, hi_res_feature=None):
...
         -------
         x : tf.Tensor
-            Output tensor with embedded hi_res_feature and observation mask
-            concatenated to input.
+            Output tensor with embedding concatenated to input.
         """
         if hi_res_feature is None:
-            hi_res_feature = tf.fill(x[..., :1].shape, np.nan)
+            hr_shape = (*x[..., 0].shape, len(self.features))
+            hi_res_feature = tf.constant(np.nan, shape=hr_shape, dtype=x.dtype)
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py:1555]: I'm confused as to why this would ever be called with `hi_res_feature=None`... Isn't the whole point of this method to integrate this input?

Also confused why you'd allow `exo_data=None if len(self.exo_features) > 0`? What would this use case be?

File: phygnn/layers/custom_layers.py [hunk 2]
```diff
@@ -1527,23 +1539,42 @@ def call(self, x, hi_res_feature=None):
         hi_res_feature : tf.Tensor | np.ndarray
             This should be a 4D array for spatial enhancement model or 5D array
             for a spatiotemporal enhancement model (obs, spatial_1, spatial_2,
-            (temporal), 1). This is NaN where there are no observations and
-            real values where observations exist.
+            (temporal), features). This is NaN where there are no observations
+            and real values where observations exist.
+        exo_data : tf.Tensor | np.ndarray
+            This is an array of exogenous data used to imform the embedding,
+            like topography
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py:1546]: Maybe specify why you'd do `hi_res_feature=None` and `exo_data=None` in the docstrings since this is default?

File: phygnn/layers/custom_layers.py [hunk 3]
```diff
@@ -1436,32 +1451,71 @@ def call(x, hi_res_feature=None):
...
+    ):
         """
         Parameters
         ----------
-        filters : int
-            Number of filters to use for each convolutional layer. e.g.
-            [64, 32] for one layer with 64 and another with 32.
         name : str | None
             Unique str identifier of the layer. Usually the name of the
             hi-resolution feature used in the concatenation.
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py:1489]: Missing docstring for "features"

File: phygnn/layers/custom_layers.py [hunk 4]
```diff
@@ -1527,23 +1539,42 @@ def call(self, x, hi_res_feature=None):
...
         if hi_res_feature is None:
-            hi_res_feature = tf.fill(x[..., :1].shape, np.nan)
+            hr_shape = (*x[..., 0].shape, len(self.features))
+            hi_res_feature = tf.constant(np.nan, shape=hr_shape, dtype=x.dtype)
+
+        if exo_data is None and len(self.exo_features) > 0:
+            exo_shape = (*x[..., 0].shape, len(self.exo_features))
+            exo_data = tf.constant(0, shape=exo_shape, dtype=x.dtype)
+
+        if self.fill_method is None:
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py:1561]: This option for `fill_method=None` is not described in the docstring yet it is supported here. Can you describe or remove? Seems like you should remove this based on the assert statement in the init

File: phygnn/layers/custom_layers.py [hunk 5]
```diff
@@ -1473,52 +1527,10 @@ def build(self, input_shape):
...
-        return tf.where(not_nan, hi_res_feature, mean), not_nan_float
-
-    def call(self, x, hi_res_feature=None):
-        """Apply the embed net to x, hi_res_feature, and the mask
-        representing where hi_res_feature is not nan. Concatenate the output
-        with x
+    def call(self, x, hi_res_feature=None, exo_data=None):
+        """Apply the embed net to x, hi_res_feature, exogenous data, and the
+        mask representing where hi_res_feature is not nan. Concatenate the
+        output with x
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py:1533]: The embed net isn't really applied to x, is it? I think the only case where x is run through the embed net is if `fill_method=None` which doesnt actually seem to be supported.

File: phygnn/layers/handlers.py [hunk 6]
```diff
@@ -278,7 +278,14 @@ def add_layer_by_class(self, class_name, **kwargs):
             logger.error(msg)
             raise KeyError(msg)
 
-        if layer_class == SkipConnection:
+        # If a layer has a hidden_layers argument it's assumed that this is
+        # given as a list of dictionaries for layers which will be initialized
+        # and then given to the parent layer.
```
Comment (grantbuster) [inline phygnn/layers/handlers.py:283]: Suggest:

```
# If a layer has a "hidden_layers" argument it's assumed that "layer_class" holds a nested set of hidden layers

File: phygnn/utilities/tf_utilities.py [hunk 7]
```diff
@@ -19,3 +20,174 @@ def tf_log10(x):
...
+    mask : tf.Tensor
+        Mask of the input tensor where True is not NaN and False is NaN.
+    """
+    not_nan_float = tf.cast(mask, x.dtype)
+
+    hi_res_zeroed = tf.where(mask, x, tf.zeros_like(x))
+
+    count = tf.reduce_sum(not_nan_float)
+    total = tf.reduce_sum(hi_res_zeroed)
+    mean = tf.math.divide_no_nan(total, count)
```
Comment (grantbuster) [inline phygnn/utilities/tf_utilities.py:52]: Maybe there's a reason you're doing this interesting mean operation but consider more intuitive syntax based on this example instead:
```
import numpy as np
import tensorflow as tf
arr = np.random.uniform(0, 1, (4, 4))
tensor = tf.convert_to_tensor(arr)
mask = tensor > 0.5
mean = tf.reduce_mean(tensor[mask])
out = tf.where(mask, tensor, mean)

File: phygnn/utilities/tf_utilities.py [hunk 8]
```diff
@@ -19,3 +20,174 @@ def tf_log10(x):
...
+    weights = 1 / dists
+    weights /= tf.reduce_sum(weights, axis=1, keepdims=True)
+
+    vals = tf.reduce_sum(weights * tf.expand_dims(valid_vals, 0), axis=1)
+
+    filled = tf.where(mask, x, tf.zeros_like(x))
+    return tf.tensor_scatter_nd_update(filled, tf.where(nan_mask), vals)
+
+
+def idw_fill(x, low_mem=True):
```
Comment (grantbuster) [inline phygnn/utilities/tf_utilities.py:118]: oh man i dont have the brainpower to give this a real math review right now, the best i can do is make sure you're aware of 

NSRDB implementation of IDW based on distances calculated from KDTree (wouldn't have to be geospatial KDTree here): https://github.com/NREL/nsrdb/blob/main/nsrdb/utilities/interpolation.py#L136

Nice OS software for IDW in tensorflow you could use: https://www.sciencedirect.com/science/article/pii/S2352711022001479
https://github.com/ElsevierSoftwareX/SOFTX-D-22-00084/blob/master/src/tfinterpy/idw.py

File: phygnn/utilities/tf_utilities.py [hunk 9]
```diff
@@ -19,3 +20,174 @@ def tf_log10(x):
...
+    """
+    nan_mask = tf.math.logical_not(mask)
+    valid_coords = tf.boolean_mask(coords, mask)
+    nan_coords = tf.boolean_mask(coords, nan_mask)
+    valid_vals = tf.boolean_mask(x, mask)
+
+    diffs = tf.expand_dims(nan_coords, 1) - tf.expand_dims(valid_coords, 0)
+    dists = tf.norm(diffs, axis=-1)
+
+    weights = 1 / dists
```
Comment (grantbuster) [inline phygnn/utilities/tf_utilities.py:109]: Any reason dist could be zero? NSRDB IDW protects against this.

**Review by grantbuster — CHANGES_REQUESTED** [review]

File: phygnn/layers/custom_layers.py (hunk 1 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py:1555]: Yeah good question. It's so I can use the model with and without these extra inputs. Train with obs and then do inference with different sets of inputs.

File: phygnn/layers/custom_layers.py (hunk 5 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py:1533]: No it's not, this is a left over from a previous setup. Thanks

File: phygnn/layers/custom_layers.py (hunk 1 above)
Comment (grantbuster) [reply phygnn/layers/custom_layers.py:1555]: Gotcha that makes sense, i think definitely include a note on this in the docstrings esp since None is the default for these args

File: phygnn/utilities/tf_utilities.py (hunk 9 above)
Comment (bnb32) [reply phygnn/utilities/tf_utilities.py:109]: No since the nan_coords and valid_coords never overlap

File: phygnn/utilities/tf_utilities.py (hunk 8 above)
Comment (bnb32) [reply phygnn/utilities/tf_utilities.py:118]: Yeah there's also the IDW in rex regridder ;). I wanted to do something that would use tensorflow methods so I avoided previous approaches with trees. This OS option seems to use that as well. I included a test that checks this against a brute force numpy approach.

Comment (bnb32) [reply phygnn/utilities/tf_utilities.py:118]: Oh the OS approach is actually pretty interesting - it sets up IDW as a tf.model with fixed layers. Most of the logic with my implementation is handling the NaNs though, and this would still be needed.

File: phygnn/utilities/tf_utilities.py (hunk 7 above)
Comment (bnb32) [reply phygnn/utilities/tf_utilities.py:52]: Yeah not totally sure why i did it this way lol

File: phygnn/layers/custom_layers.py (hunk 4 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py:1561]: Yeah this should be kept actually, so the hidden_layers = None reduces to Sup3rConcatObs.

File: phygnn/utilities/tf_utilities.py (hunk 8 above)
Comment (grantbuster) [reply phygnn/utilities/tf_utilities.py:113]: Yeah you're probably right that you can't use trees here. Sounds good just wanted to make sure you're aware of other implementations and yeah great you have tests

**Review by grantbuster — APPROVED** [review]
